Mathlib/Analysis/SpecificLimits/Basic.lean

```diff
@@ -114,6 +114,46 @@ theorem tendsto_natCast_div_add_atTop {𝕜 : Type*} [DivisionRing 𝕜] [Topolo
     intros
     simp_all only [comp_apply, map_inv₀, map_natCast]
 
+/-- If there exist real constants `b` and `B` such that for `n` big enough, `b ≤ f n ≤ B`, then
+  `f n / (n : ℝ)` tends to `0` as `n` tends to infinity. -/
+theorem tendsto_bdd_div_atTop_nhds_zero_nat {f : ℕ → ℝ} {b : ℝ}
+    (hb : ∀ᶠ n : ℕ in atTop, b ≤ f n) {B : ℝ} (hB : ∀ᶠ n : ℕ in atTop, f n ≤ B) :
+    Tendsto (fun n : ℕ => f n / (n : ℝ)) atTop (𝓝 0) := by
+  refine tendsto_of_tendsto_of_tendsto_of_le_of_le' (tendsto_const_div_atTop_nhds_zero_nat b)
+      (tendsto_const_div_atTop_nhds_zero_nat B) ?_ ?_
+  · simp only [eventually_atTop, ge_iff_le] at hb ⊢
+    obtain ⟨N, hN⟩ := hb
+    exact ⟨N, fun n hn ↦ div_le_div_of_nonneg_right (hN n hn) (cast_nonneg _)⟩
+  · simp only [eventually_atTop, ge_iff_le] at hB ⊢
+    obtain ⟨N, hN⟩ := hB
+    exact ⟨N, fun n hn ↦ div_le_div_of_nonneg_right (hN n hn) (cast_nonneg _)⟩
+
+/-- For any positive `m : ℕ`, `((n % m : ℕ) : ℝ) / (n : ℝ)` tends to `0` as `n` tends to `∞`. -/
+theorem tendsto_mod_div_atTop_nhds_zero_nat {m : ℕ} (hm : 0 < m) :
+    Tendsto (fun n : ℕ => ((n % m : ℕ) : ℝ) / (n : ℝ)) atTop (𝓝 0) := by
+  apply tendsto_bdd_div_atTop_nhds_zero_nat
+    (Eventually.of_forall (fun _ ↦ cast_nonneg _)) (B := m)
+  apply Eventually.of_forall (fun n ↦ ?_)
+  simp only [cast_le, le_of_lt (mod_lt n hm)]
+
+/-- If `u` tends to `∞` as `n` tends to `∞`, then for `n` big enough
+  `((s n : ℝ) / (u n : ℝ)) * (u n : ℝ) = (s n : ℝ)` holds. -/
+theorem div_mul_eventually_cancel (s : ℕ → ℕ) {u : ℕ → ℕ} (hu : Tendsto u atTop atTop) :
+    (fun n => (s n : ℝ) / (u n : ℝ) * (u n : ℝ)) =ᶠ[atTop] fun n => (s n : ℝ) := by
```

Comment:
There are several issues here:

1. This should be called: `Filter.EventuallyEq.div_mul_cancel`
2. It shouldn't be specialized to `atTop` or `ℕ`
3. It shouldn't be specialized to `ℝ` and make use of coercions.

Instead, it should be something like this:

```lean
theorem Filter.EventuallyEq.div_mul_cancel {α G : Type*} [GroupWithZero G] {f g : α → G}
    {l : Filter α} (hg : Tendsto g l (𝓟 {0}ᶜ)) : (fun x ↦ f x / g x * g x) =ᶠ[l] fun x ↦ f x := by
  filter_upwards [hg.le_comap <| preimage_mem_comap (m := g) (mem_principal_self {0}ᶜ)] with x hx
  aesop

theorem Filter.EventuallyEq.div_mul_cancel_atTop {α G : Type*} [LinearOrderedSemifield G]
    {f g : α → G} {l : Filter α} (hg : Tendsto g l atTop) :
    (fun x ↦ f x / g x * g x) =ᶠ[l] fun x ↦ f x :=
  div_mul_cancel <| hg.mono_right <| le_principal_iff.mpr <|
    mem_of_superset (Ioi_mem_atTop 0) <| by aesop
```

This doesn't *quite* get you what you need because of the coercions, but it's very close, and easily applied by using [tendsto_natCast_atTop_atTop](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Filter/AtTopBot/Archimedean.html#tendsto_natCast_atTop_atTop).

Does that seem reasonable to you?

Reply:
Thanks, I have replaced the previous lemma by these two.